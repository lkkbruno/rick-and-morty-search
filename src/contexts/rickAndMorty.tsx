import React, { createContext, useState, useEffect, PropsWithChildren } from 'react';
import axios from 'axios';

export interface ICharacter {
  id: number;
  name: string;
  status: string;
  species: string;
  type: string;
  gender: string;
  origin: {
    name: string;
    url: string;
  };
  location: {
    name: string;
    url: string;
  };
  image: string;
  episode: string[];
  url: string;
  created: string;
}

interface RickAndMortyContextData {
  characters: ICharacter[];
  searchTerm: string;
  setSearchTerm: (term: string) => void;
}

export const RickAndMortyContext = createContext<RickAndMortyContextData>({
  characters: [],
  searchTerm: '',
  setSearchTerm: () => { },
});

export const RickAndMortyProvider: React.FC<PropsWithChildren> = ({ children }) => {
  const [characters, setCharacters] = useState<ICharacter[]>([]);
  const [searchTerm, setSearchTerm] = useState<string>('');

  const fetchCharacters = () => {
    axios
      .get(`https://rickandmortyapi.com/api/character`, {
        params: {
          name: searchTerm,
        }
      })
      .then((response) => {
        setCharacters(response.data.results);
      }).catch(() => {
        setCharacters([]);
      });
  }

  useEffect(() => {
    fetchCharacters();
  }, [searchTerm]);

  return (
    <RickAndMortyContext.Provider value={{ characters, searchTerm, setSearchTerm }}>
      {children}
    </RickAndMortyContext.Provider>
  );
};
