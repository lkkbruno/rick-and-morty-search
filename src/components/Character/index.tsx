import React, { useContext, useState, useEffect, PropsWithChildren } from 'react';
import styled from 'styled-components';
import { RickAndMortyContext, ICharacter } from '../../contexts/rickAndMorty';

const List = styled.ul`
  display: flex;
  flex-wrap: wrap;
  justify-content: center;
  list-style: none;
  padding: 0;
  margin: 0;
`;

const Item = styled.li`
  display: flex;
  flex-direction: row;
  width: 21vw;
  height: 170px;
  background-color: rgba(208, 208, 208, 0.916);
  border-radius: 10px;
  box-shadow: 0px 0px 10px rgba(0, 0, 0, 0.1);
  padding: 0px 20px;
  margin: 20px 0px 20px 10px;

`;

const Image = styled.img`
  width: 150px;
  align-self: center;
`;

const Informations = styled.div`
padding-left: 15px;
`

const Name = styled.h2`
  color: #242424;
  margin-top: 10px;
  margin-bottom: 10px;
`;

const Information = styled.p`
  color: #242424;
  font-size: 18px;
  margin-top: 10px;
  margin-bottom: 0px;
  text-align: start;
`

interface CharacterProps {
    character: ICharacter;
}

const Character: React.FC<CharacterProps> = ({ character }) => {
    return (
        <Item key={character.id}>
            <Image src={character.image} alt={character.name} />
            <Informations>
                <Name>{character.name}</Name>
                <Information>{`Gender: ${character.gender}`}</Information>
                <Information>{`Status: ${character.status}`}</Information>
                <Information>{`Species: ${character.species}`}</Information>
            </Informations>
        </Item>
    );
};

export default Character;
