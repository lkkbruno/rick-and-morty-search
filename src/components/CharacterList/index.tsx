import React, { useContext, useState, useEffect } from 'react';
import styled from 'styled-components';
import { RickAndMortyContext, ICharacter } from '../../contexts/rickAndMorty';
import Character from '../Character';

const List = styled.ul`
  display: flex;
  flex-wrap: wrap;
  justify-content: center;
  list-style: none;
  padding: 0;
  margin: 0;
`;

interface CharacterListProps {
  characters: ICharacter[],
}

const CharacterList: React.FC<CharacterListProps> = ({ characters }) => {
  const { searchTerm } = useContext(RickAndMortyContext);
  const [filteredCharacters, setFilteredCharacters] = useState<ICharacter[]>(characters);

  useEffect(() => {
    setFilteredCharacters(
      characters.filter((character) =>
        character.name.toLowerCase().includes(searchTerm.toLowerCase())
      )
    );
  }, [characters, searchTerm]);

  return (
    <List>
      {filteredCharacters.map((character) => (
        <Character key={character.id} character={character} />
      ))}
    </List>
  );
};

export default CharacterList;
