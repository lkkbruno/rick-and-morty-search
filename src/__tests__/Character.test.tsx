import React from 'react';
import { render, waitFor, screen } from '@testing-library/react';
import CharacterList from '../components/CharacterList';
import { ICharacter, RickAndMortyProvider } from '../contexts/rickAndMorty';
import Character from '../components/Character';

const character: ICharacter = {
    id: 1,
    name: 'Ricky Sanchez Test',
    status: 'Alive',
    species: 'Human',
    type: '',
    gender: 'Male',
    origin: {
        name: 'Earth (C-137)',
        url: 'https://rickandmortyapi.com/api/location/1',
    },
    location: {
        name: 'Citadel of Ricks',
        url: 'https://rickandmortyapi.com/api/location/3',
    },
    image: 'https://rickandmortyapi.com/api/character/avatar/1.jpeg',
    episode: [
        'https://rickandmortyapi.com/api/episode/1',
    ],
    url: 'https://rickandmortyapi.com/api/character/1',
    created: '2017-11-04T18:48:46.250Z',
}


describe('CharacterList', () => {
    it('should match snapshot', async () => {
        const { container } = render(<Character character={character} />)
        expect(container).toMatchSnapshot();
    })

    it('should render a card of character and contain the text', async () => {
        render(<Character character={character} />);
        await waitFor(() => expect(screen.getByText('Ricky Sanchez Test')).toBeInTheDocument());
    });
});
