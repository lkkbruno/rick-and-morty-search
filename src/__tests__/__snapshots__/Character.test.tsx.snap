// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CharacterList should match snapshot 1`] = `
<div>
  <li
    class="sc-gEvEer cpkkRY"
  >
    <img
      alt="Ricky Sanchez Test"
      class="sc-eqUAAy cQJqLn"
      src="https://rickandmortyapi.com/api/character/avatar/1.jpeg"
    />
    <div
      class="sc-fqkvVR bFKHFo"
    >
      <h2
        class="sc-dcJsrY kWDMGg"
      >
        Ricky Sanchez Test
      </h2>
      <p
        class="sc-iGgWBj hNqgHF"
      >
        Gender: Male
      </p>
      <p
        class="sc-iGgWBj hNqgHF"
      >
        Status: Alive
      </p>
      <p
        class="sc-iGgWBj hNqgHF"
      >
        Species: Human
      </p>
    </div>
  </li>
</div>
`;
